lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,326 @@
+
+class ReportUIFeatures {
+
+  /**
+   * @param {!DOM} dom
+   */
+  constructor(dom) {
+    this.json = null;
+    /** @private {!DOM} */
+    this._dom = dom;
+    /** @private {!Document} */
+    this._document = this._dom.document();
+    /** @private {boolean} */
+    this._copyAttempt = false;
+    /** @type {Element} **/
+    this.exportButton = null;
+    /** @type {Logger} **/
+    this.logger = null;
+
+    this.onCopy = this.onCopy.bind(this);
+    this.onExportButtonClick = this.onExportButtonClick.bind(this);
+    this.onExport = this.onExport.bind(this);
+    this.onKeyDown = this.onKeyDown.bind(this);
+    this.printShortCutDetect = this.printShortCutDetect.bind(this);
+  }
+
+  /**
+   * Adds export button, print, and other functionality to the report. The method
+   * should be called whenever the report needs to be re-rendered.
+   * @param {!ReportRenderer.ReportJSON} report
+   * @param {!Element} container Parent element to render the report into.
+   */
+  initFeatures(report, container) {
+    this.json = report;
+    this._setupLogger(container);
+    this._setupExportButton();
+    this._setUpCollapseDetailsAfterPrinting();
+    this._resetUIState();
+    this._document.addEventListener('keydown', this.printShortCutDetect);
+    this._document.addEventListener('copy', this.onCopy);
+  }
+
+  /**
+   * Fires a custom DOM event on target.
+   * @param {string} name Name of the event.
+   * @param {!Document|!Element} target DOM node to fire the event on.
+   * @param {Object<{detail: Object<string, *>}>=} detail Custom data to include.
+   */
+  _fireEventOn(name, target = this._document, detail) {
+    const event = new CustomEvent(name, detail ? {detail} : null);
+    this._document.dispatchEvent(event);
+  }
+
+  _setupExportButton() {
+    this.exportButton = this._dom.find('.lh-export__button', this._document);
+    this.exportButton.addEventListener('click', this.onExportButtonClick);
+
+    const dropdown = this._dom.find('.lh-export__dropdown', this._document);
+    dropdown.addEventListener('click', this.onExport);
+  }
+
+  /**
+   * @param {!Element} container Parent element to render the report into.
+   */
+  _setupLogger(container) {
+    const loggerEl = this._dom.createElement('div');
+    loggerEl.id = 'lh-log';
+    container.appendChild(loggerEl);
+    this.logger = new Logger(loggerEl);
+  }
+
+  /**
+   * Handler copy events.
+   * @param {!Event} e
+   */
+  onCopy(e) {
+    // Only handle copy button presses (e.g. ignore the user copying page text).
+    if (this._copyAttempt) {
+      // We want to write our own data to the clipboard, not the user's text selection.
+      e.preventDefault();
+      e.clipboardData.setData('text/plain', JSON.stringify(this.json, null, 2));
+      this.logger.log('Report JSON copied to clipboard');
+    }
+
+    this._copyAttempt = false;
+  }
+
+  /**
+   * Copies the report JSON to the clipboard (if supported by the browser).
+   */
```

Comment:
I don't think there's any hope of getting `this._document.queryCommandSupported` and `this._document.execCommand` to work, so you can add `@suppress {reportUnknownTypes}` to the jsdocs of this function to get it to shut up. The only downside is any other errors in the same category in this method will be silenced (none right now)